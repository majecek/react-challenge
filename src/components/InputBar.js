import React, { Component } from 'react'

class InputBar extends Component {

  state = {
    inputTerm: ''
  }

  onInputChange = (event) => {
    console.log(event.target.value)
    this.setState({inputTerm: event.target.value})
  }

  onFormSubmit = (event) => {
    event.preventDefault()
    this.props.onInput(this.state.inputTerm.split(','))
  }

  render () {
    return (
      <div>
        <form onSubmit={this.onFormSubmit} className="input-group">
          <input
            placeholder="insert your array separated by commas"
            className="form-control"
            value={this.state.inputTerm}
            onChange={this.onInputChange}/>
          <button type="submit">Submit</button>
        </form>
      </div>
    )
  }
}

export default InputBar
