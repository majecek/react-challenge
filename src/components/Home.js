import React, { Component } from 'react'
import InputBar from './InputBar'
import ReactMatrix from 'react-matrix'
import _ from 'lodash'
import 'bootstrap/dist/css/bootstrap.min.css'

const styles = {
  container: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'center',
    padding: 10
  },

  matrix: {
    padding: 20
  }

}

class Home extends Component {

  state = {
    vectorArray: [],
    sinkBaseTotalWater: 0,
    sinkBaseArray: [],
    matrix: [[]]

  }

  onInputArray = async (vectorInput) => {
    await this.setState({
      vectorArray: vectorInput
    })

    await this.sinkBaseCalculation()
  }

  sinkBaseCalculation = () => {
    const vectorInput = this.state.vectorArray

    const leftHighestColumn = []
    for (let i = 0; i < vectorInput.length; i++) {
      leftHighestColumn[i] = Math.max(vectorInput[i], i !== 0 ? leftHighestColumn[i - 1] : 0)
    }

    const rightHighestColumn = []
    for (let i = vectorInput.length - 1; i >= 0; i--) {
      rightHighestColumn[i] = Math.max(vectorInput[i], i < vectorInput.length - 1 ? rightHighestColumn[i + 1] : 0)
    }

    const sinkBaseArray = []
    for (let i = 0; i < vectorInput.length; i++) {
      sinkBaseArray[i] = Math.min(leftHighestColumn[i], rightHighestColumn[i]) - vectorInput[i]
    }

    this.setState({
      sinkBaseTotalWater: sinkBaseArray.reduce((total, amount) => total + amount, 0),
      sinkBaseArray
    })

    this.matrixGeneration()
  }

  matrixGeneration = () => {

    const YAxis = _.max(this.state.vectorArray)
    const XAxis = this.state.vectorArray.length

    const zerosArray = []
    this.state.sinkBaseArray.forEach((waterNumber, index) => {
      zerosArray[index] = YAxis - (waterNumber + this.state.vectorArray[index])
    })

    const matrix = Array.from({length: _.max(this.state.vectorArray)}, () => [this.state.vectorArray.length].fill(0))

    for (let i = 0; i < XAxis; i++) {
      for (let j = 0; j < YAxis; j++) {
        matrix[j][i] = j < zerosArray[i] ? '0' : j < (this.state.sinkBaseArray[i] + zerosArray[i]) ? '2' : '1'
      }
    }

    this.setState({matrix})
  }

  render () {
    return (
      <div style={styles.container}>
        <div className="alert alert-info" role="alert">
          Count water in sink base - project assignment can be found on <a href="https://github.com/majecek/react-challenge">Github</a>
        </div>

        <InputBar onInput={this.onInputArray}/>

        <div className="card">
          <div className="card-block text-nowrap">
            Total water in sinkbase: {this.state.sinkBaseTotalWater}
          </div>
        </div>

        <div style={styles.matrix}>
          <ReactMatrix squareSize={50} matrix={this.state.matrix}
                       cellStates={{'0': 'available', '1': 'barrier', '2': 'path'}}/>
        </div>

      </div>
    )
  }
}

export default Home
