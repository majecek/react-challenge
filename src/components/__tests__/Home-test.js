import React from 'react'
import Home from './../Home'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import ReactMatrix from 'react-matrix'
import InputBar from './../InputBar'

it('Home component tests', () => {
  const component = renderer.create(<Home/>).toJSON()
  expect(component).toMatchSnapshot()
})

it('Home checks for sub components', () => {
  expect(shallow(<Home/>).find(InputBar)).toHaveLength(1)
  expect(shallow(<Home/>).find(ReactMatrix)).toHaveLength(1)
})

it('Testing calculations for sinkbase and matrix', async () => {
  const wrapper = shallow(<Home/>)
  // console.log(wrapper.state())

  expect(wrapper.state('sinkBaseArray')).toEqual(expect.arrayContaining([]))
  expect(wrapper.state('vectorArray')).toEqual(expect.arrayContaining([]))
  expect(wrapper.state('matrix')).toEqual(expect.arrayContaining([[]]))

  await wrapper.find(InputBar).simulate('input',[4,1,2,3])
  expect(wrapper.state('vectorArray')).toEqual(expect.arrayContaining([ 4, 1, 2, 3 ]))
  expect(wrapper.state('sinkBaseArray')).toEqual(expect.arrayContaining([ 0, 2, 1, 0 ]))
  expect(wrapper.state('matrix')).toEqual(expect.arrayContaining([[ '1', '0', '0', '0' ], [ '1', '2', '2', '1' ], [ '1', '2', '1', '1' ], [ '1', '1', '1', '1' ] ]))
  expect(wrapper.state('sinkBaseTotalWater')).toBe(3)

  await wrapper.find(InputBar).simulate('input',[1, 5, 3, 7, 2])
  // console.log(wrapper.state())
  expect(wrapper.state('sinkBaseTotalWater')).toBe(2)
  expect(wrapper.state('vectorArray')).toEqual(expect.arrayContaining([ 1, 5, 3, 7, 2 ]))
  expect(wrapper.state('sinkBaseArray')).toEqual(expect.arrayContaining([ 0, 0, 2, 0, 0 ]))
  expect(wrapper.state('matrix')).toEqual(expect.arrayContaining([[ '0', '0', '0', '1', '0' ],
    [ '0', '0', '0', '1', '0' ],
    [ '0', '1', '2', '1', '0' ],
    [ '0', '1', '2', '1', '0' ],
    [ '0', '1', '1', '1', '0' ],
    [ '0', '1', '1', '1', '1' ],
    [ '1', '1', '1', '1', '1' ] ]))

  await wrapper.find(InputBar).simulate('input',[5, 3, 7, 2, 6, 4, 5, 9, 1, 2])
  expect(wrapper.state('sinkBaseTotalWater')).toBe(14)
  await wrapper.find(InputBar).simulate('input',[5, 5, 5, 5])
  expect(wrapper.state('sinkBaseTotalWater')).toBe(0)
  await wrapper.find(InputBar).simulate('input',[5, 5, 5, 5])
  expect(wrapper.state('sinkBaseTotalWater')).toBe(0)
  // console.log(wrapper.state())
})