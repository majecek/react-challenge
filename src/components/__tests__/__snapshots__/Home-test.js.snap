// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component tests 1`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "column",
      "justifyContent": "center",
      "padding": 30,
    }
  }
>
  <div>
    <form
      className="input-group"
      onSubmit={[Function]}
    >
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="insert your array separated by comma"
        value=""
      />
      <span
        className="input-group-btn"
      >
        <button
          className="btn btn-secondary"
          type="submit"
        >
          Submit
        </button>
      </span>
    </form>
  </div>
  <div
    className="card"
  >
    <div
      className="card-block text-nowrap"
    >
      Total water in sinkbase: 
      0
    </div>
  </div>
  <div>
    <svg
      className="Matrix"
      height={50}
      width={0}
    >
      <g>
        <g
          className="Row"
        />
      </g>
    </svg>
  </div>
</div>
`;
